pyfstat/snr.py

```diff
@@ -0,0 +1,420 @@
+from attrs import define, field
+import lal
+import lalpulsar
+import logging
+import numpy as np
+from typing import Union
+
+from pyfstat.helper_functions import get_ephemeris_files
+
+
+@define(kw_only=True, slots=False)
+class SignalToNoiseRatio:
+    """
+    Compute the optimal signal-to-noise ratio (and derived quantities)
+    of a CW signal on Gaussian noise or an arbitrary dataset of SFTs.
+
+    The definition of SNR (shortcut for "optimal signal-to-noise ratio")
+    is taken from Eq. (76) of https://dcc.ligo.org/LIGO-T0900149 and is
+    such that `<2F> = 4 + SNR^2`, where `<2F>` represents the expected
+    value over noise realization of twice the F-statistic of a prefectly
+    matched template to an existing signal in the data.
+
+    Computing SNR^2 requires two quantities:
+        - The M matrix, which depends on the sky position and polarization angle
+        and encodes the effect of the detector's antenna pattern over the course
+        of the observing run.
+        - The JKS amplitude paramaters {A^0, A^1, A^2, A^3}, which are functions
+        of the CW's amplitude parameters (h0, cosi, psi) or, alternatively,
+        (aPlus, aCross, psi).
+
+    Parameters
+    ----------
+    detector_states: lalpulsar.MultiDetectorStateSeries
+        MultiDetectorStateSeries as produced by DetectorStates.
+        Provides the required information to compute the antenna pattern contribution.
+    noise_weights: Union[lalpulsar.MultiNoiseWeights, None]
+        Optional, incompatible with `(assumeSqrtSX, Tsft)`.
+        Can be compute from SFTs using `SignalToNoiseRatio.from_sfts`.
+        Vector of noise weights to account for a varying noise floor or unequal noise
+        floors in different detectors.
+    assumeSqrtSX: float
+        Optional, incompatible with `noise_weights`.
+        *Single-sided* *amplitude* spectral density (ASD) of the detector noise.
+        This value is used for all detectors, meaning it's not possible to manually
+        specify different noise floors without creating SFT files.
```

Comment:
Deeper problems: I cannot access MultiNoiseWeights from Python as there's no constructor.

PFS can do it because for different sqrtSX it sets up the weights manually, but that cannot be done without proper access to MultiNoiseWeights. It's simple once updated in LALSuite though.

Reply:
Ah, that's where that comment was referring to. Ok to keep for future upgrade, then.